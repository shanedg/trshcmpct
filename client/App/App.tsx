import React from 'react';

const App = () => {
  return (
    <article>

      <header>
        <h1>trshcmpctr.com</h1>
      </header>

      <section>
        <article>
          <h2>welcome</h2>
          <p>hey hello hi</p>

          <section>
            <h3>the highlights</h3>
            <ul>
              <li>list</li>
              <li>item</li>
            </ul>
          </section>

        </article>
      </section>

    </article>
  );
};

export default App;
