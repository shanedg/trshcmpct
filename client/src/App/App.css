a {
  color: aqua;
}

a:visited {
  color: blueviolet;
}

.navigation-list {
  list-style-type: none;
  display: inline-flex;
  padding: 0;
}

.navigation-list li {
  padding-right: 6px;
}
