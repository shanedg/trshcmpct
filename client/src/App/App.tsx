import React, { StrictMode } from 'react';
import { createBrowserRouter, RouterProvider, Link, useParams, Outlet } from 'react-router-dom';

import './App.css';
import { useGuildMemberData } from './use-guild-member-data';
import { Welcome } from './Welcome';

type GuildUserData = ReturnType<typeof useGuildMemberData>;

const getWelcomeMessage = (guildUser: GuildUserData) => {
  const welcome = 'welcome to the trash compactor';

  if (guildUser?.user?.username) {
    return `${welcome}, ${guildUser.user.username}`;
  }

  return `${welcome}, <unknown>`;
};

const Home = () => {
  const guildUser = useGuildMemberData();

  if (guildUser) {
    return (
      <>
        <ul className="navigation-list">
          <li>
            <Link to="/new">new</Link>
          </li>
          <li>
            <Link to="/worlds">worlds</Link>
          </li>
        </ul>
        <Welcome message={getWelcomeMessage(guildUser)} />
      </>
    );
  }

  return (<p>loading...</p>);
};

const WorldsList = () => {
  const fakeWorlds = [
    {
      id: 1,
      label: 'world one',
      version: '1.16.5',
      createdAt: '2023/06/28',
      lastOnline: '2023/06/28',
      createdBy: '@shaned.gg'
    },
    {
      id: 2,
      label: 'world two',
      version: '1.19.0',
      createdAt: '2023/06/28',
      lastOnline: '2023/06/28',
      createdBy: '@shaned.gg'
    },
    {
      id: 3,
      label: 'world three',
      version: '1.20.1',
      createdAt: '2023/06/28',
      lastOnline: '2023/06/28',
      createdBy: '@shaned.gg'
    },
  ];

  return (
    <>
      <ul className="navigation-list">
        <li>
          <Link to="/">back</Link>
        </li>
      </ul>
      <table>
        <thead>
          <tr>
            <td>name</td>
            <td>version</td>
            <td>created</td>
            <td>last online</td>
            <td>created by</td>
          </tr>
        </thead>
        <tbody>
          {fakeWorlds.map(({ id, label, version, createdAt, createdBy, lastOnline }) => {
            return (
              <tr key={`${id}-${label}-${version}`}>
                <td><Link to={`${id}`}>{label}</Link></td>
                <td>{version}</td>
                <td>{new Date(createdAt).toLocaleDateString()}</td>
                <td>{new Date(lastOnline).toLocaleDateString()}</td>
                <td>{createdBy}</td>
              </tr>
            );
          })}
        </tbody>
      </table>
      <Outlet />
    </>
  );
};

const WorldDetail = () => {
  const { worldId } = useParams();
  return (
    <>
      <h2>{worldId}</h2>
    </>
  );
};

const NewWorld = () => {
  return (
    <>
      <ul className="navigation-list">
        <li>
          <Link to="/">back</Link>
        </li>
      </ul>
      <form>
        <label htmlFor="version">version</label>
        <select name="version">
          <option>1.20.1</option>
        </select>
        <label htmlFor="name">name</label>
        <input name="name" />
        <button>create</button>
      </form>
    </>
  );
};

const router = createBrowserRouter([
  {
    path: '/',
    element: <Home />,
  },
  {
    path: '/worlds',
    element: <WorldsList />,
    children: [
      {
        path: ':worldId',
        element: <WorldDetail />,
      },
    ],
  },
  {
    path: '/new',
    element: <NewWorld />,
  },
]);

const App = () => {
  return (
    <StrictMode>
      <h1>trshcmpctr</h1>
      <RouterProvider router={router} />
    </StrictMode>
  );
};

export default App;
