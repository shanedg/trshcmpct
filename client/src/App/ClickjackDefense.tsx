import React, { FC, useEffect } from 'react';

/**
 * Server state rendered to html for comparison with state saved to local storage
 */
interface ClickjackDefenseServerState {
  state: string;
}

/**
 * Declare server->client window data namespace
 * i.e.
 * window.__trshcmpctr = { ... }
 */
declare global {
  interface Window { __trshcmpctr__: ClickjackDefenseServerState }
}

/**
 * Invisible component that throws an error if the server state doesn't match local storage
 * Helps ensure that authentication was initiated by our application
 */
export const ClickjackDefense: FC = () => {
  useEffect(() => {
    // render template state encoded to avoid unintended html character escaping
    const templateState: string = window.atob(decodeURIComponent(window.__trshcmpctr__.state));
    const localStorageState = localStorage.getItem('oauth-state');

    if (localStorageState === null) {
      throw new Error(`you may have been clickjacked!
  local storage state is null`);
    }

    if (templateState !== localStorageState) {
      throw new Error(`you may have been clickjacked!
  server state: ${templateState}
  local storage state: ${localStorageState}`);
    }
  }, []);
  return <></>;
};
