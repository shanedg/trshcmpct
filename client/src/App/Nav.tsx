import React, {
  FC,
  PropsWithChildren,
} from 'react';

interface NavLinkMeta {
  href: string;
  label?: string;
}

// Treat link label as children
type NavLinkProps = PropsWithChildren<Pick<NavLinkMeta, 'href'>>;

const NavLink: FC<NavLinkProps> = ({ href, children }) => (
  <li>
    <a href={href}>
      {children || href}
    </a>
  </li>
);

/**
 * Minimal navigation link list
 */
export const Nav: FC<{ links: NavLinkMeta[] }> = ({ links }) => (
  <nav>
    <ul>
      {links.map(({ href, label }, i) => (
        <NavLink key={`${label || href}--${i}`} href={href}>
          {label}
        </NavLink>
      ))}
    </ul>
  </nav>
);
