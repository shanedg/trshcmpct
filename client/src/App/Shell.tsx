import React from 'react';

import type { FC, PropsWithChildren } from 'react';

export const Shell: FC<PropsWithChildren> = ({ children }) => {
  return (<>{children}</>);
};
