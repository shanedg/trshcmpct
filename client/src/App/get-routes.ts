import { createElement } from 'react';

import { ErrorCard } from './components/ErrorCard';
import { Home } from './components/Home';
import { NavigationChrome } from './components/NavigationChrome';
import { NewWorld } from './components/NewWorld';
import { WorldDetail } from './components/WorldDetail';
import { Worlds } from './components/Worlds';

import type { RouteObject } from 'react-router';

type NamedRoute = Pick<RouteObject, 'element' | 'errorElement'>;

type NamedRouteWithChildren = NamedRoute & {
  children?: Record<string, NamedRouteWithChildren>;
}

export type NamedRoutes = Record<string, NamedRouteWithChildren>;

/**
 * A map of named routes declaring the application route hierarchy
 * Not understood by react-router browser router
 */
export const namedRoutes: NamedRoutes = {
  '/': {
    children: {
      '/': {
        element: createElement(Home),
        errorElement: createElement(ErrorCard, { error: Error('Problem rendering Home')}),
      },
      '/worlds': {
        children: {
          ':worldId': {
            element: createElement(WorldDetail),
            errorElement: createElement(ErrorCard, { error: Error('Problem rendering World')}),
          },
        },
        element: createElement(Worlds),
        errorElement: createElement(ErrorCard, { error: Error('Problem rendering Worlds')}),
      },
      '/new': {
        element: createElement(NewWorld),
        errorElement: createElement(ErrorCard, { error: Error('Problem rendering New')}),
      },
    },
    element: createElement(NavigationChrome),
    errorElement: createElement(ErrorCard, { error: Error('Problem rendering routes')}),
  },
};

/**
 * Transform named routes map into a list of routes for react-router browser router
 * @param routes Named routes
 * @returns A list of routes for browser router
 */
export function transformRoutesForBrowserRouter(routes: NamedRoutes): RouteObject[] {
  const childRoutes: RouteObject[] = [];
  for (const childPath in routes) {
    const { children, ...namedRoute } = routes[childPath];
    if (children) {
      childRoutes.push({
        ...namedRoute,
        children: transformRoutesForBrowserRouter(children),
        path: childPath,
      });
      // Skip the following push because it's already handled
      continue;
    }
    // Push a children-less version of the route if not handled above
    childRoutes.push({
      ...namedRoute,
      path: childPath,
    });
  }
  return childRoutes;
}

/**
 * Get app routes
 * @returns A list of routes for browser router
 */
export const getRoutes = () => transformRoutesForBrowserRouter(namedRoutes);
