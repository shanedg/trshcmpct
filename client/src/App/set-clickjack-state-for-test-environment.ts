/**
 * This module has side effects on the test environment:
 * - sets `window.__trshcmpctr__` property (normally set by server)
 * - stubs `window.localStorage` for jsdom
 */

/**
 * Mock implementation of browser local storage
 */
class LocalStorageMock {
  store: Record<string, string> = {};

  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: string) {
    this.store[key] = String(value);
  }

  removeItem(key: string) {
    delete this.store[key];
  }

  // stub
  get length() { return 1; }

  // stub
  key(_index: number): string | null { return 'some-key-name'; }
}

/**
 * Helper for exercising and bypassing clickjack defenses in the test environment
 */
export const setClickjackStateForTestEnvironment = (
  templateState = 'some-state',
  // local storage can return null if no item with key
  localStorageState: string | null = 'some-state'
) => {
  // template state is encoded to avoid unintended character escaping when rendered to html
  const uriEncodedBase64State = encodeURIComponent(window.btoa(templateState));

  const localStorageMock = new LocalStorageMock();
  // don't bother storing null for an item, just don't call setItem
  if (localStorageState !== null) {
    localStorageMock.setItem('oauth-state', localStorageState);
  }

  window.__trshcmpctr__ = { state: uriEncodedBase64State };

  Object.defineProperty(window, 'localStorage', {
    value: localStorageMock,
    writable: true,
  });
  window.localStorage = localStorageMock;
};

/**
 * Helper that cleans up module changes to the test environment
 */
export const cleanupClickjackStateForTestEnvironment = () => {
  // @ts-expect-error: window.__trshcmpctr__ isn't optional
  delete window.__trshcmpctr__;
  // @ts-expect-error: window.localStorage isn't optional
  delete window.localStorage;
};
