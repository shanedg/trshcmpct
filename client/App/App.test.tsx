import React from 'react';

import '@testing-library/jest-dom/extend-expect';
import {
  render,
  screen,
} from '@testing-library/react';

import App from './App';

describe('App', () => {
  it('renders', async () => {
    render(<App />);

    await screen.findByRole('heading');

    expect(screen.getByRole('heading')).toHaveTextContent('trshcmpctr');
  });
});
