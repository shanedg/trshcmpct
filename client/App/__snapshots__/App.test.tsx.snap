// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `"trshcmpctr.comwelcomehey hello hithe highlightslistitem"`;
